# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/wasm/WasmIonCompile.cpp
# Commit: bba9e8259714
# Full Hash: bba9e82597147f0a5b9a6a6a8fc6525ff87fc914
# Author: Dmitry Bezhetskov <dbezhetskov@igalia.com>
# Date: 2020-10-02 04:10:47
# Regressor Bug: 1639153
# File Overlap Count: 1
# Description:
#   Bug 1639153 - Part 6.3: Establish dependency from tls for arm callWithABI div/mod i64. r=lth
#   
#   To be able to call c++ runtime via Builtin thunk we need to set up WasmTlsReg.
#   In this patch I create dependencies from MIR level to Codegen to be sure that WasmTlsReg is alive
#   when we call runtime in div/mod i64 for arm.
# ==============================================================================

diff -r c5ef1152da2b -r bba9e8259714 js/src/wasm/WasmIonCompile.cpp
--- a/js/src/wasm/WasmIonCompile.cpp	Thu Oct 01 07:52:24 2020 +0000
+++ b/js/src/wasm/WasmIonCompile.cpp	Thu Oct 01 07:52:37 2020 +0000
@@ -461,9 +461,9 @@
       rhs = rhs2;
     }
 
-    // For x86 we implement i64 div via c++ runtime.
-    // A call to c++ runtime requires tls pointer.
-#ifdef JS_CODEGEN_X86
+    // For x86 and arm we implement i64 div via c++ builtin.
+    // A call to c++ builtin requires tls pointer.
+#if defined(JS_CODEGEN_X86) || defined(JS_CODEGEN_ARM)
     if (type == MIRType::Int64) {
       auto* ins =
           MWasmBuiltinDivI64::New(alloc(), lhs, rhs, tlsPointer_, unsignd,
@@ -495,9 +495,9 @@
       rhs = rhs2;
     }
 
-    // This is because x86 codegen calls runtime via BuiltinThunk and so it
-    // needs WasmTlsReg to be live.
-#ifdef JS_CODEGEN_X86
+    // For x86 and arm we implement i64 mod via c++ builtin.
+    // A call to c++ builtin requires tls pointer.
+#if defined(JS_CODEGEN_X86) || defined(JS_CODEGEN_ARM)
     if (type == MIRType::Int64) {
       auto* ins =
           MWasmBuiltinModI64::New(alloc(), lhs, rhs, tlsPointer_, unsignd,
